_posts/2017-11-16-Performance-of-LSTM-Using-CuDNN-v5.md

```diff
@@ -0,0 +1,136 @@
+
+When we use a large LSTM setting (layer=3, length=50), about 7 times faster in forward time, and 4 times faster in backward time with cuDNN.
+
+When we use a small LSTM setting (layer=1, length=5), about 1.5 times faster in forward time, and 1.04 times faster in backward time.
+
+
+![The effect of the layer size of LSTM and sequence length of data]({{ site.url }}/assets/nsteplstm-cudnn-v5-length-layers.png)
+
+* The effect of the layer size of LSTM and random sequence length of data
+parameters: layer = {1, 2, 3}, random={True, False}
+In this setting, we compare if the data sequence length is fixed or not. (Max length is 25.)
+When we use cuDNN, the effect of randomness is small. 
+
+![The effect of the layer size of LSTM and random sequence length of data]({{ site.url }}/assets/nsteplstm-cudnn-v5-random-seq.png)
+
+* The effect of the layer size of LSTM and the input unit size
+parameters:  layer = {1, 2, 3}, input={128, 256}
+In the setting with cuDNN, as number of layers increases, the difference between cuDNN and no-cuDNN will be large. (About 5 times faster in forward time, and about 2.7 times faster in backward time.)
+
+
+![The effect of the layer size of LSTM and the input unit size]({{ site.url }}/assets/nsteplstm-cudnn-v5-input.png)
+
+
+
+
+* The effect of the layer size of LSTM and the hidden unit size
+parameters: layer = {1, 2, 3}, hidden={128, 256}
+In the setting with cuDNN, the number of layers increases, the difference between cuDNN and no-cuDNN will be large. (This is same as layer size and input size experiment.)
+However as hidden unit size increases, the difference between cuDNN and no-cuDNN will be small. 
+
+![The effect of the layer size of LSTM and the hidden unit size]({{ site.url }}/assets/nsteplstm-cudnn-v5-hidden.png)
+
+
+* The effect of the layer size of LSTM and dropout rate
+parameters: layer={1, 2, 3}, dropout={0.5, 0.0}
+In the setting with cuDNN, when using dropout, the speed gets slower but the difference is very small (dropout rate=0.50).
+
+![The effect of the layer size of LSTM and dropout rate]({{ site.url }}/assets/nsteplstm-cudnn-v5-drop.png)
+
+
+
+* When will the differences be large? (Setting with/without cuDNN.)
+
+As batch size is small (batchsize=128), sequence length is long (length=50) and the number of layers is large (layer=3), the difference is large. (The cuDNN is faster than no-cuDNN setting.)
+
+If we use a large LSTM in the experiment, the effect of cuDNN will be large.
+7.8 times faster in forward time.
+4.0 times faster in backward time.
+
+ 
+
+
+### Experimental Environment
+* GPU: GeForce GTX 970
+* Chainer (v1.21)
+* cuDNN v5.1 (cuda v8)
+ 
+### Experimental Setting
+* Data: Random artificial sequence data (data size: 10,000)
+* Training epoch: 10
+* Comparing the average time per one epoch.
+* Performance time of
+  * forward time (for train data)
+  * forward time (for test data)
+  * backward time
+  * Default experiment setting:
+  * batchsize : 128
```

Comment:
This list of settings (L122-127) looks further indented or marked in the original draft.